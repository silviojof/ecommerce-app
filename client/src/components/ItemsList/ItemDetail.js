import React from 'react';
import ic_shipping from '../../images/ic_shipping.png';
import ic_shipping2x from '../../images/ic_shipping@2x.png';
import Mobile1 from '../../images/mobile1.jpg';

import styles from './Items.module.scss';

const ItemDetail = ({ data }) => {
    return (
        <article className={styles.article}>
            <a href={`/items/${data.id}`} className={styles.articleGrid}>
                <div className={styles.imageContainer}>
                    <img src={Mobile1} alt="product" />
                </div>
                <div className={styles.infoContainer}>
                    <h3>
                        {`${data.price.currency} ${data.price.amount}`}
                        {
                            data.free_shipping &&
                            <img src={ic_shipping} srcSet={`${ic_shipping2x} 2x`} alt="shipping status" />
                        }
                    </h3>
                    <p>{data.title}</p>
                </div>
                <div className={styles.locationContainer}>
                    <h6>{data.location}</h6>
                </div>
            </a>
            <hr />
        </article>
    )
};

export default ItemDetail;