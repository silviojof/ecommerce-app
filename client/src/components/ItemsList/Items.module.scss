@import '../../styles/colors';

.container {
    @extend %container;
}

.products {
    @extend %paper;
}

.article {
    .articleGrid {
        text-decoration: none;
        color: inherit;
        outline: 0;
        cursor: pointer;
        display: grid;
        grid-template-columns: 180px 1fr 180px;
        padding: $spacing1;
    }
    hr {
        width: 95%;
        height: 1px;
        background-color: $lightGrey;
        border: none;
    }
    &:last-of-type {
        hr {
            display: none;
        }
    }
}

.imageContainer {
    width: 180px;
    height: 180px;
    img {
        border-radius: $radius;
        height: 100%;
        width: 100%;
        object-fit: cover;
    }
}

.infoContainer {
    padding: $spacing1;
    h3 {
        margin-bottom: $spacing2;
        font-weight: 500;
        margin-top: 0;
        display: flex;
        align-items: center;
        font-size: 24px;
        img {
            margin-left: 10px;
        }
    }
    p {
        font-size: 18px;
        margin: 0;
    }
}

.locationContainer {
    padding: $spacing1; 
    h6 {
        font-size: 12px;
        font-weight: 400;
        color: $mediumGrey;
        margin: 0;
    }
}