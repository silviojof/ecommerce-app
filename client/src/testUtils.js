import React from 'react';
import { render } from 'react-testing-library';
import { Provider } from 'react-redux';
import createSagaMiddleware from 'redux-saga';
import configureStore from 'redux-mock-store';

const sagaMiddleware = createSagaMiddleware()
const middlewares = [sagaMiddleware]

export const mockStore = configureStore(middlewares)

export const renderWithProvider = ({ store, component } = {}) => {
  return render(
    <Provider store={store}>
        {component}
    </Provider>
  )
};