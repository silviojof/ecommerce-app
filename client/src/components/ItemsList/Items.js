import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchItems, clearItems } from '../../redux/ducks/productList';
import ItemDetail from './ItemDetail';
import Breadcrumb from '../Breadcrumb';
import Loader from '../Loader';

import styles from './Items.module.scss';

const Items = ({ location, fetchItems, productList, clearItems, isLoading }) => {
    useEffect(() => {
        const searchParams = new URLSearchParams(location.search);
        const search = searchParams.get('search');
        if (search) {
            fetchItems(search)
        }
        return () => {
            clearItems();
        };
      }, [location.search]);

    if (isLoading) {
        return (
            <div className={styles.loader}>
                <Loader />
            </div>
        )
    }

    return (
            <div className={styles.container}>
                <Breadcrumb />
                {
                    productList.length > 0 ?
                    <section className={styles.products}>
                        {
                            productList.map(el => (
                                <ItemDetail key={el.id} data={el} />
                            ))
                        }
                    </section>
                    :
                    <div>
                        No hay elementos dentro de este criterio de b√∫squeda.
                    </div>
                }
            </div>
    )
};

Items.propTypes = {
    location: PropTypes.object.isRequired,
    fetchItems: PropTypes.func.isRequired,
    productList: PropTypes.array,
    clearItems: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired,
};

Items.defaultProps = {
    productList: []
};

const mapStateToProps = (state) => {
    return {
        productList: state.productList.items,
        isLoading: state.productList.isLoadingItems
    }
}

const mapDispatchToProps = dispatch => {
    return {
      fetchItems: args => dispatch(fetchItems(args)),
      clearItems: () => dispatch(clearItems())
    }
  }

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(Items);