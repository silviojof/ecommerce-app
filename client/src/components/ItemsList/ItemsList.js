import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { fetchItems, clearItems } from '../../redux/ducks/productList';
import ItemDetail from './ItemDetail';
import Breadcrumb from '../Breadcrumb';
import Loader from '../Loader';

import styles from './Items.module.scss';

export const ItemsList = ({ location, fetchItems, productList, clearItems, isLoading }) => {
    useEffect(() => {
        const searchParams = new URLSearchParams(location.search);
        const search = searchParams.get('search');
        if (search) {
            fetchItems(search)
        }
        return () => {
            clearItems();
        };
      }, [location.search]);

    if (isLoading) {
        return (
            <div data-testid="loader" className={styles.loader}>
                <Loader />
            </div>
        )
    }

    return (
            <div className={styles.container}>
                <Breadcrumb />
                {
                    productList.length > 0 ?
                    <section data-testid="articles" className={styles.products}>
                        {
                            productList.map(el => (
                                <ItemDetail key={el.id} data={el} />
                            ))
                        }
                    </section>
                    :
                    <div>
                        No hay elementos dentro de este criterio de b√∫squeda.
                    </div>
                }
            </div>
    )
};

ItemsList.propTypes = {
    location: PropTypes.object.isRequired,
    fetchItems: PropTypes.func.isRequired,
    productList: PropTypes.arrayOf(PropTypes.shape({
        id: PropTypes.string.isRequired,
        location: PropTypes.string.isRequired,
        picture: PropTypes.string.isRequired,
        condition: PropTypes.string.isRequired,
        free_shipping: PropTypes.bool.isRequired,
        price: PropTypes.shape({
            currency: PropTypes.string.isRequired,
            amount: PropTypes.number.isRequired,
            decimals: PropTypes.number.isRequired,
        }),
    })).isRequired,
    clearItems: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => {
    return {
        productList: state.productList.items,
        isLoading: state.productList.isLoadingItems
    }
}

const mapDispatchToProps = dispatch => {
    return {
      fetchItems: args => dispatch(fetchItems(args)),
      clearItems: () => dispatch(clearItems())
    }
  }

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ItemsList);