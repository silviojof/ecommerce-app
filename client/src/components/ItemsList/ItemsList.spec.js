import React from 'react';
import {render} from 'react-testing-library';
import { renderWithProvider, mockStore } from '../../testUtils';
import { ItemsList } from './ItemsList';

const defaultProps = {
    location: {},
    match: {
        params: {}
    },
    fetchItems: jest.fn(),
    productList: [{
        id: 'MLA23',
        title: 'title',
        location: 'Buenos Aires',
        picture: '',
        condition: 'new',
        free_shipping: true,
        price: {
            currency: 'ARS',
            amount: 23,
            decimals: 12,
        },
    }],
    clearItems: jest.fn(),
    isLoading: false,
};

const initialStateStore = mockStore({ productList: { categories: ['first', 'second']} });

it ('matches snapshot', () => {
    const { container } = renderWithProvider({
        store: initialStateStore,
        component: (
            <ItemsList {...defaultProps} />
        )
    });
    expect(container.firstChild).toMatchSnapshot();
});

it('isLoading', () => {
    const newProps = {
        ...defaultProps,
        isLoading: true
    };
    const {getByTestId} = render(<ItemsList {...newProps} />);
    expect(getByTestId('loader')).toBeInTheDocument();
});

it('no Products', () => {
    const newProps = {
        ...defaultProps,
        productList: [],
    };
    const { queryByText } = renderWithProvider({
        store: initialStateStore,
        component: (
            <ItemsList {...newProps} />
        )
    });
    expect(queryByText('No hay elementos dentro de este criterio de bÃºsqueda.')).toBeInTheDocument();
});

it('load Products', () => {
    const newProps = {
        ...defaultProps
    };
    const { getByTestId } = renderWithProvider({
        store: initialStateStore,
        component: (
            <ItemsList {...newProps} />
        )
    });
    expect(getByTestId('articles')).toBeInTheDocument();
});