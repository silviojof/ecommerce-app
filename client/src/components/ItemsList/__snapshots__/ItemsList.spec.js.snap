// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="container"
>
  <div
    class="breadcrumb"
  >
    <div
      class="breadcrumbItem"
    >
      <span
        class=""
      >
        first
      </span>
      <span
        class="breadcrumbArrow"
      >
        &gt;
      </span>
    </div>
    <div
      class="breadcrumbItem"
    >
      <span
        class="last"
      >
        second
      </span>
    </div>
  </div>
  <section
    class="products"
    data-testid="articles"
  >
    <article
      class="article"
    >
      <a
        class="articleGrid"
        href="/items/MLA23"
      >
        <div
          class="imageContainer"
        >
          <img
            alt="product"
            src=""
          />
        </div>
        <div
          class="infoContainer"
        >
          <h3>
            $ 23
            <img
              alt="shipping status"
              src="ic_shipping.png"
              srcset="ic_shipping@2x.png 2x"
            />
          </h3>
          <p>
            title
          </p>
        </div>
        <div
          class="locationContainer"
        >
          <h6>
            Buenos Aires
          </h6>
        </div>
      </a>
      <hr />
    </article>
  </section>
</div>
`;
